javascript/ql/test/library-tests/TypeScript/Types/tst.ts

```diff
@@ -293,4 +293,83 @@ module TS46 {
       payload.toFixed(); // <- number
     }
   };
-}
\ No newline at end of file
+}
+
+const key = Symbol();
+
+const numberOrString = Math.random() < 0.5 ? 42 : "hello";
+
+let obj = {
+  [key]: numberOrString,
+};
+
+if (typeof obj[key] === "string") {
+  let str = obj[key]; // <- string
+  str.toUpperCase();
+}
+
+//////////
+
+function f<T>(arg: {
+  produce: (n: string) => T,
+  consume: (x: T) => void }
+): void {};
+
+f({
+  produce: n => n, // <- (n: string) => string
+  consume: x => x.toLowerCase()
+});
+
+///////////
+
+const ErrorMap = Map<string, Error>;
+
+const errorMap = new ErrorMap(); // <- Map<string, Error>
+
+////////////
+
+type FirstString<T> =
+  T extends [infer S extends string, ...unknown[]]
```

Comment:
Do we extract the `extends string` part?

Usually we've tried to extract enough of the AST to prevent spurious unused variable/import alerts. Since the `extends string` part could instead contain a reference to an imported name, we'd want to extract it.

The unused variable alert isn't that important anymore, but I still think we should try to maintain it.

Reply:
Yes we extract it, and I've added a test that it works with our unused variable query.